--- a/python/triton/runtime/build.py
+++ b/python/triton/runtime/build.py
@@ -93,6 +93,14 @@ def _build(name, src, srcdir, library_dirs, include_dirs, libraries):
         if system == "Linux" and machine in ("aarch64", "arm64"):
             # On Arm backend, some CPU (neoverse-v2) needs to be specified through -mcpu
             cc_cmd += ["-mcpu=native"]
+
+    # Nixpkgs support branch
+    # Allows passing e.g. extra -Wl,-rpath
+    cc_cmd_extra_flags = "@ccCmdExtraFlags@"
+    if cc_cmd_extra_flags != ("@" + "ccCmdExtraFlags@"): # substituteAll hack
+        import shlex
+        cc_cmd.extend(shlex.split(cc_cmd_extra_flags))
+
     ret = subprocess.check_call(cc_cmd)
     if ret == 0:
         return so
